### ISSUE TITLE
Service controller reconciles static node set causing UpdateLoadBalancerFailed events

### ISSUE THREAD

**Issue by maxwolffe** [kubernetes/kubernetes#109180]
### What happened?

We're managing our own Kubernetes cluster on Azure. During node refreshes (in which we're aiming to update the node pool with new VMs with updated VHDs) - we add several new nodes to the cluster, select several old nodes to remove, drain those nodes, and then delete the underlying VM (part of a VMSS) and the old node one at a time. During deletion of these old nodes, we frequently see `UpdateLoadBalancerFailed` events of the following form:

```
UpdateLoadBalancerFailed - Error updating load balancer with new hosts map[new-node-1:{} new-node-2:{} new-node-3:{} old-node-2:{} old-node-3:{} 
```

The UpdateLoadBalancer event is interesting because it includes some nodes which we had just removed - it appears to be attempting to update the load balancer with nodes which have already been deleted. 

Digging into the service_controller logs, we see the following events (edited slightly):
```
Mar 25, 2022 @ 19:49:00 - service_controller.go:690] Detected change in list of current cluster nodes. New node set: map[new-node-1:{} new-node-2:{} new-node-3:{} old-node-2:{} old-node-3:{}]
Mar 25, 2022 @ 19:52:22 - service_controller.go:771] Finished syncing service "production/good0".
Mar 25, 2022 @ 19:56:20 - service_controller.go:771] Finished syncing service "production/good1".
Mar 25, 2022 @ 19:56:02 - service_controller.go:714] failed to update load balancer hosts for service production/xx-x0: EnsureHostInPool(production/xx-x0): backendPoolID(/subscriptions/yy-yy/resourceGroups/cindiaprod/providers/Microsoft.Network/loadBalancers/kubernetes-standard/backendAddressPools/kubernetes) - failed to ensure host in pool: "instance not found"
Mar 25, 2022 @ 20:01:38 - service_controller.go:714] failed to update load balancer hosts for service production/xx-x1: EnsureHostInPool(production/xx-x1): backendPoolID(/subscriptions/yy-yy/resourceGroups/cindiaprod/providers/Microsoft.Network/loadBalancers/kubernetes-standard/backendAddressPools/kubernetes) - failed to ensure host in pool: "instance not found"
Mar 25, 2022 @ 20:04:43 - service_controller.go:714] failed to update load balancer hosts for service production/xx-x2: EnsureHostInPool(production/xx-x2): backendPoolID(/subscriptions/yy-yy/resourceGroups/cindiaprod/providers/Microsoft.Network/loadBalancers/kubernetes-standard/backendAddressPools/kubernetes) - failed to ensure host in pool: "instance not found"
...
Mar 25, 2022 @ 21:03:40.967 - Successfully updated 2 out of 19 load balancers to direct traffic to the updated set of nodes
Mar 25, 2022 @ 21:03:40.967  service_controller.go:690] Detected change in list of current cluster nodes: map[new-node-1:{} new-node-2:{} new-node-3:{}]
```

After the first set of failures, the update succeeds without issue, and is referencing only the new set of nodes.

**Our cluster state looks something like the following:**
LoadBalancer Services - ~20
Nodes in cluster - ~90
Pods -  ~1200

**Hypothesis**
We noticed that service_controller.go is fetching the list of nodes once - https://github.com/kubernetes/kubernetes/blob/806723e6ec622c30b86caf69ddd76871ceb187bf/pkg/controller/service/service_controller.go#L690 - and then using that to update the many load balancer services in the cluster. Each service takes 3-4 minutes to update, during which time the set of nodes which are actually running in the cluster may have changed. When EnsureHostInPool https://github.com/kubernetes/kubernetes/blob/806723e6ec622c30b86caf69ddd76871ceb187bf/staging/src/k8s.io/legacy-cloud-providers/azure/azure_vmss.go#L1091 encounters a node which is no longer in the VMSS, it returns an "instance not found" error, which is the cause of our UpdateLoadBalancer failure events. 

We were surprised that the service reconciliation logic doesn't handle the underlying node set changing (since syncing many services can take a long time) 

** Workaround ** 
In our case, we create our VMSS template with the load_balancer_backend_address_pools_ids field already set, so the  new VMs which we're adding to our cluster are already members of the LoadBalancer's backend pool. The VMs which we're deleting have been removed from the cluster, so we don't expect them to be included in the set of nodes to be used to update the LoadBalancer's backend pool. 

Because the service_controller isn't performing actions on the VM which seems necessary (we may be misunderstanding this) - we've added the `kubernetes.azure.com/managed=false` label to our nodes, so they are not included in the load balancer update (excluded in this block - https://github.com/kubernetes/kubernetes/blob/806723e6ec622c30b86caf69ddd76871ceb187bf/staging/src/k8s.io/legacy-cloud-providers/azure/azure_vmss.go#L1101) - this has stopped the errors without seeming to impact routing to the nodes (they're included in the load balancer backend pool)

** Other notes ** 
- This cluster is running an older version of kubernetes - so this may have been fixed in newer releases. 
- Happy to provider more information on our set up if it would be helpful. 

### What did you expect to happen?

We did not expect to see UpdateLoadBalancerFailed events during node deletion.

### How can we reproduce it (as minimally and precisely as possible)?

We can reproduce this consistently on our clusters running k8s 1.16.12 by shutting down a VM in our VMSS and deleting the associated k8s node, waiting for 5 minutes, and then doing that again. 

### Anything else we need to know?

@yangl900 has context from our end and suggested that @feiskyer would be a good point of context from the Kubernetes team. 

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.4", GitCommit:"3cce4a82b44f032d0cd1a1790e6d2f5a55d20aae", GitTreeState:"clean", BuildDate:"2021-08-11T18:16:05Z", GoVersion:"go1.16.7", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.12", GitCommit:"17c50ce2d686f4346924935063e3a431360e0db7", GitTreeState:"clean", BuildDate:"2020-06-26T03:33:27Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
WARNING: version difference between client (1.21) and server (1.16) exceeds the supported minor version skew of +/-1
```

</details>

### Cloud provider

<details>
Azure
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by yangl900** [comment]
@feiskyer what do you think? Does it make sense for service controller to refresh the node list before it sends to cloud provider to reconcile service? Or, should the azure provider EnsureHostsInPool ignores the stale, has been deleted nodes?

**Comment by feiskyer** [comment]
thanks @maxwolffe and @yangl900 raising the issue. I think it makes sense for `EnsureHostsInPool()` to ignore the deleted nodes instead of refreshing node list. Note that node deletion may have some delays after VM deletion, hence even if node list refreshed, the same issue may continue to happen.

**Comment by feiskyer** [comment]
@nilo19 let's get this fixed

/assign @nilo19

**Comment by nilo19** [comment]
We have a fix for this. https://github.com/kubernetes/kubernetes/pull/105188. But it's not in your version v1.16.12. Can you switch to a newer version that has the fix? @maxwolffe

**Comment by maxwolffe** [comment]
Thanks friends! Looks like the oldest version which has this fix is 1.23? 

We're a little ways off from that update, but it looks like the issue is resolved (from reading the PR), feel free to close this if needed.

**Comment by bridgetkromhout** [comment]
Hi, @maxwolffe - okay, as the fix exists in 1.23 and beyond, I'll close this issue now. (And I'll encourage you to get your updates accomplished, as 1.16.x no longer receives security patches. If it turns out that AKS or CAPZ meets your needs, that might help you stay more up-to-date. Thanks!)

/close
